### PR TITLE
Add code linting

### LINKED ISSUE #2: Add syntax checking
Can syntax validation be enabled? It would be very helpful for users when editing PHP, JS, or CSS to know whether or not it has syntax errors. 

Once that is active, on the theme and plugin editor screens, we should prevent saving the file if there are syntax errors. Likewise, for custom CSS in the customizer we should mark the setting as invalid to prevent it from being published.

See also https://core.trac.wordpress.org/ticket/39218

### PR DESCRIPTION
**Pull request by georgestephanis**
Fixes #2 

Doesn't offer PHP syntax checking -- we may need to do some sort of ajax command to have the server lint the php code, I'm not seeing much by way of pure JS syntax checking.

### CODE REVIEW

**Comment by georgestephanis** [comment]
The error tooltips don't seem to be displaying for me in the customizer for some reason -- if anyone can take a look at why, that'd be appreciated.

**Comment by georgestephanis** [comment]
Tooltips now show in the customizer -- I had to bump their z-index to `500000` because of reasons.

File: codemirror-wp.php [hunk 1]
```diff
@@ -25,6 +25,18 @@ public static function register_scripts() {
 		wp_register_script( 'codemirror-addon-hint-sql',        plugins_url( "CodeMirror/addon/hint/sql-hint.js", __FILE__ ),        array( 'codemirror-addon-hint-show', 'codemirror-mode-sql' ), SELF::CODEMIRROR_VERSION );
 		wp_register_script( 'codemirror-addon-hint-xml',        plugins_url( "CodeMirror/addon/hint/xml-hint.js", __FILE__ ),        array( 'codemirror-addon-hint-show', 'codemirror-mode-xml' ), SELF::CODEMIRROR_VERSION );
 
+		// The linting engines for the lint addons...
+		wp_register_script( 'csslint',  'http://csslint.net/js/csslint.js' );
+		wp_register_script( 'htmlhint', 'http://htmlhint.com/js/htmlhint.js' );
+		wp_register_script( 'jshint',   '//ajax.aspnetcdn.com/ajax/jshint/r07/jshint.js' );
+		wp_register_script( 'jsonlint', 'https://rawgithub.com/zaach/jsonlint/79b553fb65c192add9066da64043458981b3972b/lib/jsonlint.js' );
```
Comment (westonruter) [inline codemirror-wp.php:32]: Eventually these will need to be bundled.

File: codemirror-wp.php [hunk 2]
```diff
@@ -250,6 +275,9 @@ public static function customize_controls_print_footer_scripts() {
 		#customize-control-custom_css .CodeMirror {
 			height: calc( 100vh - 185px );
 		}
+		.CodeMirror-lint-tooltip {
+			z-index: 500000;
```
Comment (westonruter) [inline codemirror-wp.php:279]: The reason for this is https://github.com/WordPress/wordpress-develop/blob/0b5b1a667f28c76cbc9db3d2a16ced99353f6f77/src/wp-admin/css/themes.css#L1323

File: codemirror-wp.php (hunk 1 above)
Comment (georgestephanis) [reply codemirror-wp.php:32]: Agreed.  I'm planning on setting them as externals before actually merging this.  Or maybe just importing wholesale.  Unsure.

**Comment by westonruter** [comment]
So there is no PHP syntax error checking implemented in JS for CodeMirror? What about http://codemirror.net/1/contrib/php/ ?

**Comment by georgestephanis** [comment]
Hadn't seen that previously, I'm playing with it now.  Probably because I was only looking at the CM Core repo.

Thanks!

**Comment by georgestephanis** [comment]
Ah, that's for the old CodeMirror 1 branch -- current is v5-- I'm exploring to see what's feasible.

**Comment by georgestephanis** [comment]
I'm sure there's some optimizations involving promises that I can have ^^ do, but for the moment I can't even get the helper to run to do the linting -- I'm not sure what I'm doing wrong, maybe fresh eyes can help?

**Comment by westonruter** [comment]
It makes me really nervous allow arbitrary PHP to be written to the file system, even though it's in a temporary directory. I am dubious as to whether the `php` binary will be available, and I think `shell_exec` is disabled by many hosts for security reasons. If PHP is to be used for linting, then I think the `php_check_syntax()` function should be used instead, but alas it is deprecated was removed from PHP. I don't see a good server-side option here.

**Comment by georgestephanis** [comment]
I agree completely.

Unfortunately I haven't found a pure JS linter for PHP -- they're all build
tool wrappers around a PHP binary.

**Comment by georgestephanis** [comment]
I'm nuking my work on a PHP linter -- I'm really not happy with it, hopefully we'll find something better if we do anything at all.

**Pull request merged**